### PR TITLE
Withdraw tests

### PR DESCRIPTION
**Pull request by bachstatter**
- Add `InsufficientCollateral` error to avoid panic underflow weh users try to withdraw more than avail collateral
- Add some more params to the `MaxCollateralExceeded`
- Add test case for `InsufficientCollateral`
- Add test case for  `InsufficientCollateralAvailableForWithdraw`
- Add test case for  `AccountLiquidatable` 
- Add test case to make sure we can withdraw with open positions

For now the tests for `InsufficientCollateralAvailableForWithdraw` and `AccountLiquidatable` are fairly simple. There's lots of permutations that could be tested, but I think more scenarios are covered in the liquidation tests. And I think it make sense to keep it like this.. 

There is one more case we could write a tests for, and that is making sure we cant withdraw when an account is flagged, will tackle that in a separate PR. 

Let me know if there's good reason to keep errors lean with less params.

### CODE REVIEW

File: markets/perps-market/contracts/storage/GlobalPerpsMarket.sol [hunk 1]
```diff
@@ -47,16 +53,28 @@ library GlobalPerpsMarket {
...
+                    synthAmount.toUint()
+                );
             } else {
                 self.collateralAmounts[synthMarketId] += synthAmount.toUint();
             }
         } else {
+            uint synthAmountAbs = MathUtil.abs(synthAmount);
+            if (collateralAmount < synthAmountAbs) {
+                revert InsufficientCollateral(synthMarketId, collateralAmount, synthAmountAbs);
+            }
```
Comment (sunnyvempati) [inline markets/perps-market/contracts/storage/GlobalPerpsMarket.sol:76]: makes sense

File: markets/perps-market/test/integration/Account/ModifyCollateral.withdraw.test.ts [hunk 2]
```diff
@@ -114,4 +117,155 @@ describe('ModifyCollateral Withdraw', () => {
...
+        );
+      });
+    });
+    describe('withdraw with open position', () => {
+      it('allow withdraw when its less than "collateral available for withdraw"', async () => {
+        before('withdraw allowed amount', async () => {
+          await systems()
+            .PerpsMarket.connect(trader1())
+            .modifyCollateral(trader1AccountId, sUSDSynthId, bn(-17000));
+        });
```
Comment (sunnyvempati) [inline markets/perps-market/test/integration/Account/ModifyCollateral.withdraw.test.ts:227]: this is odd; you have the `before` inside the `it` block.  surprised the test passed

**Review by sunnyvempati — COMMENTED** [review]
have one change but pre approving.

File: markets/perps-market/test/integration/Account/ModifyCollateral.withdraw.test.ts (hunk 2 above)
Comment (bachstatter) [reply markets/perps-market/test/integration/Account/ModifyCollateral.withdraw.test.ts:227]: Opsie, will fix

File: markets/perps-market/contracts/storage/GlobalPerpsMarket.sol (hunk 1 above)
Comment (bachstatter) [reply markets/perps-market/contracts/storage/GlobalPerpsMarket.sol:76]: I noticed we have `InsufficientMargin` in the AsyncOrder file. Do we have a definition on the differences between margin and collateral? And do we want to refuse errors as much as possible?

File: markets/perps-market/test/integration/Account/ModifyCollateral.withdraw.test.ts (hunk 2 above)
Comment (bachstatter) [reply markets/perps-market/test/integration/Account/ModifyCollateral.withdraw.test.ts]: Fixed!

File: markets/perps-market/contracts/storage/GlobalPerpsMarket.sol (hunk 1 above)
Comment (bachstatter) [reply markets/perps-market/contracts/storage/GlobalPerpsMarket.sol:76]: Hmm we also have a `error InsufficientMarginError(uint leftover);` in PerpsAccount.

Comment (sunnyvempati) [reply markets/perps-market/contracts/storage/GlobalPerpsMarket.sol:76]: we should distinguish them in this way imo:

- `InvalidWithdrawalAmount(amountAvailableToWithdraw, amountRequested)`: this one for the above scenario where trader is attempting to withdraw more than the entire market has.
- `InsufficientMargin()`: currently this is thrown when validating orders.  I have a TODO to create separate errors for which this is being reused during validation.
- `InsufficientCollateral()`: this should occur very rarely cuz by the time you're deducting, you've already checked you have enough to cover fees.  so yeah lets rename `InsufficientMarginError` to this or something similar?

all these are suggestions; feel free to suggest other names and ways to distinguish 👍🏽

**Review by sunnyvempati — APPROVED** [review]

**Pull request merged**